src/python/pants/util/memo.py

```diff
@@ -4,8 +4,18 @@
 import functools
 import inspect
 from contextlib import contextmanager
+from typing import Any, Callable, Optional, TypeVar
 
-from pants.util.meta import classproperty, staticproperty
+from pants.util.meta import P, classproperty, staticproperty
+
+
+# NB: We liberally ignore MyPy issues in the implementations of these decorators. This is because
+#  it would be hard to both type the implementation correctly and get the call sites of the
+#  decorators properly typed. We care far more about the latter use case.
+
+
+FuncType = Callable[..., Any]
```

Comment:
Is it possible to make this a parameterized type instead of using `Any` here?

Reply:
There's no need to, as far as I can tell. I took these two lines from MyPy docs: 

https://mypy.readthedocs.io/en/stable/generics.html#declaring-decorators